clerk.go

```diff
@@ -0,0 +1,360 @@
+		RawJSON:    body,
+	}
+}
+
+// APIRequest describes requests to the Clerk API.
+type APIRequest struct {
+	Method string
+	Path   string
+	Params Queryable
+}
+
+// SetParams sets the APIRequest.Params.
+func (req *APIRequest) SetParams(params Queryable) {
+	req.Params = params
+}
+
+// NewAPIRequest creates an APIRequest with the provided HTTP method
+// and path.
+func NewAPIRequest(method, path string) *APIRequest {
+	return &APIRequest{
+		Method: method,
+		Path:   path,
+	}
+}
+
+// Backend is the primary interface for communicating with the Clerk
+// API.
+type Backend interface {
+	// Call makes requests to the Clerk API.
+	Call(context.Context, *APIRequest, ResponseReader) error
+}
+
+// ResponseReader reads Clerk API responses.
+type ResponseReader interface {
+	Read(*APIResponse)
+}
+
+// Queryable can add parameters to url.Values.
+// Useful for constructing a request query string.
+type Queryable interface {
+	Add(url.Values)
+}
+
+// BackendConfig is used to configure a new Clerk Backend.
+type BackendConfig struct {
+	// HTTPClient is an HTTP client instance that will be used for
+	// making API requests.
+	// If it's not set a default HTTP client will be used.
+	HTTPClient *http.Client
+	// URL is the base URL to use for API endpoints.
+	// If it's not set, the default value for the Backend will be used.
+	URL *string
+}
+
+// NewBackend returns a default backend implementation with the
+// provided configuration.
+// Please note that the return type is an interface because the
+// Backend is not supposed to be used directly.
+func NewBackend(config *BackendConfig) Backend {
+	if config.HTTPClient == nil {
+		config.HTTPClient = httpClient
+	}
+	if config.URL == nil {
+		config.URL = String(APIURL)
+	}
+	return &defaultBackend{
+		HTTPClient: config.HTTPClient,
+		URL:        *config.URL,
+	}
+}
+
+// GetBackend returns the library's supported backend for the Clerk
+// API.
+func GetBackend() Backend {
+	var b Backend
+
+	backend.mu.RLock()
+	b = backend.Backend
+	backend.mu.RUnlock()
+
+	if b != nil {
+		return b
+	}
+
+	b = NewBackend(&BackendConfig{})
+	SetBackend(b)
+	return b
+}
+
+// SetBackend sets the Backend that will be used to make requests
+// to the Clerk API.
+// Use this method if you need to override the default Backend
+// configuration.
+func SetBackend(b Backend) {
+	backend.mu.Lock()
+	defer backend.mu.Unlock()
+	backend.Backend = b
+}
+
+type defaultBackend struct {
+	HTTPClient *http.Client
+	URL        string
+}
+
+// Call sends requests to the Clerk API and handles the responses.
+func (b *defaultBackend) Call(ctx context.Context, apiReq *APIRequest, setter ResponseReader) error {
+	req, err := b.newRequest(ctx, apiReq)
+	if err != nil {
+		return err
+	}
+
+	return b.do(req, apiReq.Params, setter)
+}
+
+func (b *defaultBackend) newRequest(ctx context.Context, apiReq *APIRequest) (*http.Request, error) {
+	path := b.URL + apiReq.Path
+	req, err := http.NewRequest(apiReq.Method, path, nil)
```

Comment:
🔧 I think you can pass the context here directly: https://pkg.go.dev/net/http#NewRequestWithContext